File: fashionMNISTClassifier.py
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical


class FashionMNISTClassifier:
    def __init__(self):
        self.model = None
        self.input_shape = (28, 28, 1)

    # Load and preprocess the Fashion MNIST dataset
    def load_and_preprocess_data(self):
        # Load Fashion MNIST dataset
        (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

        # Normalize images
        train_images = train_images.astype('float32') / 255.0
        test_images = test_images.astype('float32') / 255.0

        # Reshape for CNN input
        train_images = train_images.reshape((-1, 28, 28, 1))
        test_images = test_images.reshape((-1, 28, 28, 1))

        # One-hot encode labels
        train_labels = to_categorical(train_labels, 10)
        test_labels = to_categorical(test_labels, 10)

        return train_images, train_labels, test_images, test_labels

    # Build the CNN model
    def build_model(self):

        # Define the CNN model with 6 layers
        self.model = models.Sequential([
            # 1st layer: Conv
            layers.Conv2D(32, (3, 3), activation='relu', input_shape=self.input_shape),
            layers.MaxPooling2D((2, 2)),

            # 2nd layer: Conv
            layers.Conv2D(64, (3, 3), activation='relu'),
            layers.MaxPooling2D((2, 2)),

            # 3rd layer: Conv
            layers.Conv2D(64, (3, 3), activation='relu'),

            # 4th layer: Flatten
            layers.Flatten(),

            # 5th layer: Dense
            layers.Dense(64, activation='relu'),

            # 6th layer: Output
            layers.Dense(10, activation='softmax')
        ])

        # Compile the model
        self.model.compile(optimizer='adam',
                            loss='categorical_crossentropy',
                            metrics=['accuracy'])

    # train the model    
    def train(self, train_images, train_labels, epochs=5):
        self.model.fit(train_images, train_labels, epochs=epochs, batch_size=64, verbose=2)

    # Evaluate the model on test data
    def evaluate(self, test_images, test_labels):
        test_loss, test_acc = self.model.evaluate(test_images, test_labels, verbose=0)
        print(f"Test Accuracy: {test_acc:.4f}")

    # Predict and visualize some images
    def predict_images(self, images, labels, class_names, count=2):
        predictions = self.model.predict(images[:count])

        for i in range(count):
            plt.imshow(images[i].reshape(28, 28), cmap='gray')
            plt.title(f"Actual: {class_names[np.argmax(labels[i])]} | Predicted: {class_names[np.argmax(predictions[i])]}")
            plt.axis('off')
            plt.show()
    
if __name__ == "__main__":
    classifier = FashionMNISTClassifier()
    train_images, train_labels, test_images, test_labels = classifier.load_and_preprocess_data()
    classifier.build_model()
    # Train the model
    classifier.train(train_images, train_labels, epochs=5)
    classifier.evaluate(test_images, test_labels)

    class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat','Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
    # Predict 2 sample images
    classifier.predict_images(test_images, test_labels, class_names, count=2)
  
    

File: fashionMNISTClassifier.R
# Install required packages if not already done
if (!require("keras")) install.packages("keras")
if (!require("R6")) install.packages("R6")

library(keras)
library(R6)

# Install TensorFlow backend if not already installed
#keras::install_keras()  # Uncomment if running for the first time

CNNClassifier <- R6Class("CNNClassifier",
  public = list(
    model = NULL,
    train_images = NULL,
    train_labels = NULL,
    test_images = NULL,
    test_labels = NULL,
    
    initialize = function() {
      cat("Initializing CNN Classifier...\n")
      self$load_data()
      self$preprocess_data()
      self$build_model()
    },
    
    load_data = function() {
      cat("Loading Fashion MNIST dataset...\n")
      data <- dataset_fashion_mnist()
      self$train_images <- data$train$x
      self$train_labels <- data$train$y
      self$test_images <- data$test$x
      self$test_labels <- data$test$y
    },
    
    preprocess_data = function() {
      cat("Preprocessing data...\n")
      self$train_images <- self$train_images / 255
      self$test_images <- self$test_images / 255
      
      self$train_images <- array_reshape(self$train_images, c(nrow(self$train_images), 28, 28, 1))
      self$test_images <- array_reshape(self$test_images, c(nrow(self$test_images), 28, 28, 1))
      
      self$train_labels <- to_categorical(self$train_labels, 10)
      self$test_labels <- to_categorical(self$test_labels, 10)
    },
    
    build_model = function() {
      cat("Building CNN model with 6 layers...\n")
      self$model <- keras_model_sequential() %>%
        layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = 'relu', input_shape = c(28, 28, 1)) %>%
        layer_max_pooling_2d(pool_size = c(2, 2)) %>%
        layer_conv_2d(filters = 64, kernel_size = c(3, 3), activation = 'relu') %>%
        layer_max_pooling_2d(pool_size = c(2, 2)) %>%
        layer_flatten() %>%
        layer_dense(units = 64, activation = 'relu') %>%
        layer_dense(units = 10, activation = 'softmax')
      
      self$model %>% compile(
        loss = 'categorical_crossentropy',
        optimizer = optimizer_adam(),
        metrics = 'accuracy'
      )
    },
    
    train_model = function(epochs = 5, batch_size = 64) {
      cat("Training model...\n")
      self$model %>% fit(
        self$train_images, self$train_labels,
        epochs = epochs,
        batch_size = batch_size,
        validation_split = 0.2
      )
    },
    
    evaluate_model = function() {
      cat("Evaluating model on test data...\n")
      score <- self$model %>% evaluate(self$test_images, self$test_labels)
      #cat("Test accuracy:", score$accuracy, "\n")
      cat("Test accuracy:", score[["accuracy"]])
    },
    
    predict_image = function(index = 1) {
        cat("Making prediction on test image", index, "...\n")
        #dim(self$test_images)
        image <- self$test_images[index,,,,drop = FALSE]
        prediction <- self$model %>% predict(image)
        print(prediction)
        cat("Predicted class:", which.max(prediction) - 1, "\n")  # zero-indexed
    }
  )
)

cnn <- CNNClassifier$new()
cnn$train_model(epochs = 5)
cnn$evaluate_model()
cnn$predict_image(index = 1)
cnn$predict_image(index = 2)